pkgs/top-level/all-packages.nix

```diff
@@ -838,18 +838,22 @@ with pkgs;
             fetchurl = stdenv.fetchurlBoot;
           };
         });
-        perl = buildPackages.perl.override { fetchurl = stdenv.fetchurlBoot; };
+        perl = buildPackages.perl.override { fetchurl = stdenv.fetchurlBoot; inherit zlib; };
         openssl = buildPackages.openssl.override {
           fetchurl = stdenv.fetchurlBoot;
           buildPackages = {
-            coreutils = buildPackages.coreutils.override {
+            coreutils = (buildPackages.coreutils.override rec {
               fetchurl = stdenv.fetchurlBoot;
               inherit perl;
               xz = buildPackages.xz.override { fetchurl = stdenv.fetchurlBoot; };
               gmp = null;
               aclSupport = false;
               attrSupport = false;
-            };
+              autoreconfHook = null;
+              texinfo = null;
+            }).overrideAttrs (_: {
+              preBuild = "touch Makefile.in"; # avoid automake
+            });
```

Comment:
I think this change should not be needed if your overrides don't trigger a rebuild of `fetchurl`. If you do need to rebuild `fetchurl`I would suggest handling in the rest of overrides (added a separate longer comment as an alternative not to rebuild `fetchpatch`). WDYT?

Reply:
As someone who just used this patch: I'd defer to the simpler user-facing API (i.e. not asking the user to predicate behavior on whether or not `fetchurl` needed to be rebuilt).

Reply:
I don't worry about the rebuilds. The problem happened here because the overlay introduced a dependency cycle which was not there initially. The change is a workaround for a specific case of newly introduced loop.

It's a workaround I don't feel good about: conditionally disabling `autoreconf` on `coreutils` is not great. It would be easier to accept the change if it was a change that strips (or adds) some dependencies unconditionally.

If others think this change is reasonable to break this particular loop type then so be it. I still think it's the overlay user's responsibility to untangle new dependency cycles they introduce.

And to state the obvious: i'm not an authority on `stdenv` (or anything in `nixpkgs`).

Reply:
Oh, I misunderstood. It seems like you're saying that tthis change to `fetchurl` is necessary _because_ of the "solution". It's not that an unchanged `fetchurl` was the _source_ of the original infinite recursion. Okay, yeah, I ran into this issue and tried using this branch and I actually _still_ saw failure to build `cpio` (even though it solved the infinite recursion failure) - `nix log` is attached below.

[6akz7axh5y7393qm0a8w4n5nwkb0wjvp-cpio-2.13.drv.log](https://github.com/user-attachments/files/17967776/6akz7axh5y7393qm0a8w4n5nwkb0wjvp-cpio-2.13.drv.log)

Okay, I agree that if my (new) understanding is right that this isn't the best way to go. I have a few follow-up questions, since you seem to have a pretty deep working understanding of `bootstrap` and how `stdenv`'s are defined.

1. I initially tried to use this work as-is (point `nixpkgs` to this fork) but ran into an issue building `cpio` (see above). So, then I tried to use this patch on `24.11` and ran into an infinite recursion. How did @dguibert track down those packages which were related to infinite recursions? Or, how would someone, generally?
```bash
$ nix build .\#defaultLibcxxStdenv --show-trace --override-input nixpkgs $HOME/code/repos/others/nixos/nixpkgs              
warning: Git tree '/home/john/code/repos/mine/f-u' is dirty
warning: Git tree '/home/john/code/repos/others/nixos/nixpkgs' is dirty
warning: not writing modified lock file of flake 'git+file:///home/john/code/repos/mine/f-u':
• Updated input 'nixpkgs':
    'github:dguibert/nixpkgs/e2adbbf71a6d6e49ecf2f66e48acdbab2dccdaa3' (2022-11-18)
  → 'git+file:///home/john/code/repos/others/nixos/nixpkgs' (2024-11-28)
trace: Enabling Custom Stdenv
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'f-u++'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildInputs' of derivation 'f-u++'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:383:7:

          382|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          383|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          384|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       … while evaluating derivation 'nix-2.24.10'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildInputs' of derivation 'nix-2.24.10'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:383:7:

          382|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          383|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          384|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       … while evaluating derivation 'brotli-1.1.0'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'nativeBuildInputs' of derivation 'brotli-1.1.0'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:380:7:

          379|       depsBuildBuild              = elemAt (elemAt dependencies 0) 0;
          380|       nativeBuildInputs           = elemAt (elemAt dependencies 0) 1;
             |       ^
          381|       depsBuildTarget             = elemAt (elemAt dependencies 0) 2;

       … while evaluating derivation 'cmake-3.30.5'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildInputs' of derivation 'cmake-3.30.5'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:383:7:

          382|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          383|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          384|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       … while evaluating derivation 'libarchive-3.7.7'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildInputs' of derivation 'libarchive-3.7.7'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:383:7:

          382|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          383|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          384|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       … while evaluating derivation 'e2fsprogs-1.47.1'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildInputs' of derivation 'e2fsprogs-1.47.1'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:383:7:

          382|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          383|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          384|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       … while evaluating derivation 'fuse-3.16.2'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'nativeBuildInputs' of derivation 'fuse-3.16.2'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:380:7:

          379|       depsBuildBuild              = elemAt (elemAt dependencies 0) 0;
          380|       nativeBuildInputs           = elemAt (elemAt dependencies 0) 1;
             |       ^
          381|       depsBuildTarget             = elemAt (elemAt dependencies 0) 2;

       … while evaluating derivation 'meson-1.6.0'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildInputs' of derivation 'meson-1.6.0'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:383:7:

          382|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          383|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          384|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       … while evaluating derivation 'openmp-18.1.8'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildInputs' of derivation 'openmp-18.1.8'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:383:7:

          382|       depsHostHost                = elemAt (elemAt dependencies 1) 0;
          383|       buildInputs                 = elemAt (elemAt dependencies 1) 1;
             |       ^
          384|       depsTargetTarget            = elemAt (elemAt dependencies 2) 0;

       … while evaluating derivation 'llvm-18.1.8'
         whose name attribute is located at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'nativeBuildInputs' of derivation 'llvm-18.1.8'

         at /nix/store/93ixbis1xdzcq6jjbnkc2kv72qsgg5h8-source/pkgs/stdenv/generic/make-derivation.nix:380:7:

          379|       depsBuildBuild              = elemAt (elemAt dependencies 0) 0;
          380|       nativeBuildInputs           = elemAt (elemAt dependencies 0) 1;
             |       ^
          381|       depsBuildTarget             = elemAt (elemAt dependencies 0) 2;

       error: infinite recursion encountered

       at «none»:0: (source not available)
```

2. In @dguibert's minimal reproduction he did more than just pull in this patch, he also redefines `glibc`, `binutils`, and `gcc` and then does some magic with them:
```nix
    replaceStdenv = { pkgs }: let
      # Bootstrap a new stdenv that includes a modified glibc
      glibc = pkgs.glibc.overrideDerivation (old: {
        postInstall = old.postInstall + ''
          touch $out/MODIFIED_GLIBC
        '';
      });
      binutils = pkgs.binutils.override {
        libc = glibc;
      };
      gcc = pkgs.gcc.override {
        bintools = binutils;
        libc = glibc;
      };
    in
     builtins.trace "Enabling Custom Stdenv" pkgs.stdenv.override {
      cc = gcc;
      overrides = self: super: {
        inherit glibc binutils gcc;
      };
      allowedRequisites = pkgs.stdenv.allowedRequisites ++
        [ glibc.out glibc.dev glibc.bin binutils ];
    };
 ```
I found this was necessary on top of `24.11` (at least, it alleviated the infinite recursion at eval-time - it remains to be seen if `cpio` is in any better state). Do you think something like this is generally necessary, or did I just "luck out" in that it fixed my eval issue?

3. How would you recommend solving this problem, instead?